### ISSUE TITLE
Central-dashboard doesn't sync iframe url to document when navigation happens programmatically

### ISSUE THREAD

**Issue by Bobgy** [kubeflow/kubeflow#4598]
/kind bug

This is a corner case of https://github.com/kubeflow/kubeflow/issues/3042

**What steps did you take and what happened:**
1. Use pipelines UI
2. Clone an existing run
3. It goes to experiment page with run list
4. Refresh page

Bug: it goes back to clone run page
Expected: it should stay in experiment page

**Anything else you would like to add:**
Env: in kubeflow deployment (I verified KFP standalone doesn't have this bug)

When the page is in the bug state, location of KFP iframe is https://multiuser-bugbash.endpoints.gongyuan-pipeline-test.cloud.goog/pipeline/#/experiments/details/a984b224-e675-4dae-960f-ba0254d0940a, but the root page is https://multiuser-bugbash.endpoints.gongyuan-pipeline-test.cloud.goog/_/pipeline/?ns=kubeflow-gongyuan#/runs/new?cloneFromRun=88b132f6-f414-48ce-b223-3ffab1e3b271

The path mismatch is the problem. Somehow centraldashboard didn't update this navigation.

I did some investigation and found the root cause:
Central dashboard detects iframe url change every time a click happens and syncs it to root url: https://github.com/kubeflow/kubeflow/blob/master/components/centraldashboard/public/components/iframe-container.js#L54
and https://github.com/kubeflow/kubeflow/blob/master/components/centraldashboard/public/components/main-page.js#L264

However, in this bug, the navigation happens in this sequence:
1. user clicks create button
2. a network request is sent to create the run
3. when the request returns successfully, navigate to run list page

Therefore, the navigation hasn't happened when page was clicked, so it isn't captured by central dashboard.

I think it would be too complex if central dashboard monitors every type of event that might trigger a navigation.
Two choices:
* Also monitors for https://developer.mozilla.org/en-US/docs/Web/API/Window/hashchange_event of the iframe. This should work with clients using hash navigation.
* Add a timer which checks url updates every second. (In cases other than click, I think 1 second delay of syncing back the url is good enough)

What do you think about this?

**Environment:**

- Kubeflow version: (version number can be found at the bottom left corner of the Kubeflow dashboard): 0.7.0

**Comment by Bobgy** [comment]
@prodonjs @avdaredevil 
Ping after holiday. Any thoughts?

**Comment by prodonjs** [comment]
@Bobgy - Thanks for the excellent analysis!

/cc @kwasi who is now working on Kubeflow front-end projects with @avdaredevil.

TL;DR - Sounds like either the `hashchange` trigger or polling after a click are viable options and would be good for quick fixes. Long term, it makes sense to investigate better integration between the applications and centraldashboard by extending the event handling library if team cycles permit.

This is definitely an edge case that the currently implementation fails to provide an adequate solution for. I think your solution of adding some sort of polling for a navigation change after a click event would work. I would imagine that we'd want to place some sort of upper bound on the number of polls  (maybe 5s?). However, if it's really just Pipelines (and possibly Metadata) that we're concerned about, and both of them use hash-based routing, perhaps the `hashchange` event is the better option.

We did write a relatively simple library that could be included in the iframed-application code (https://github.com/kubeflow/kubeflow/blob/master/components/centraldashboard/public/library.js). However, that would require more extensive work on both the Central Dashboard and Pipelines codebases to build the more sophisticated integration. I do think that in the long run, doing something like this would be advantageous rather than finding these different edge cases and fixing them as one offs. However, I'm not really involved in the KF front-end work at this time so I can't speak for what the plans are for the year and how this might fit in.

**Comment by Bobgy** [comment]
Thanks @prodonjs for shedding some light on this.

To clarify, I didn't mean polling after click. Just periodic polling sounds better. This event handler is very lightweight, I don't think it will cause any problems.

**Comment by Bobgy** [comment]
@avdaredevil any chance this can be dealt with soon?
I can submit a PR too if you are busy.

**Comment by avdaredevil** [comment]
Hey @Bobgy, I don't have cycles till mid-Feb to tackle any more Kubeflow task items (since I'm heads down on MLMD work currently). I'm more than happy to review any PRs you send my way though :)

**Comment by Bobgy** [comment]
No problem, I will submit a PR for this.
